x-pack/plugins/beats/server/routes/api/register_assign_tags_to_beats_route.js

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Joi from 'joi';
+import {
+  get,
+  flatten,
+  uniq
+} from 'lodash';
+import { INDEX_NAMES } from '../../../common/constants';
+import { callWithRequestFactory } from '../../lib/client';
+import { wrapEsError } from '../../lib/error_wrappers';
+
+async function getDocs(callWithRequest, ids) {
+  const params = {
+    index: INDEX_NAMES.BEATS,
+    type: '_doc',
+    body: { ids },
+    _source: false
+  };
+
+  const response = await callWithRequest('mget', params);
+  return get(response, 'docs', []);
+}
+
+function getBeats(callWithRequest, beatIds) {
+  const ids = beatIds.map(beatId => `beat:${beatId}`);
+  return getDocs(callWithRequest, ids);
+}
+
+function getTags(callWithRequest, tags) {
+  const ids = tags.map(tag => `tag:${tag}`);
+  return getDocs(callWithRequest, ids);
+}
+
+async function findNonExistentItems(callWithRequest, items, getFn) {
+  const itemsFromEs = await getFn.call(null, callWithRequest, items);
+  return itemsFromEs.reduce((nonExistentItems, itemFromEs, idx) => {
+    if (!itemFromEs.found) {
+      nonExistentItems.push(items[idx]);
+    }
+    return nonExistentItems;
+  }, []);
+}
+
+function findNonExistentBeatIds(callWithRequest, beatIds) {
+  return findNonExistentItems(callWithRequest, beatIds, getBeats);
+}
+
+function findNonExistentTags(callWithRequest, tags) {
+  return findNonExistentItems(callWithRequest, tags, getTags);
+}
+
+async function persistAssignments(callWithRequest, assignments) {
+  const body = flatten(assignments.map(assignment => {
+    const { beatId, tag } = assignment;
+    const script = 'def beat = ctx._source.beat; '
+      + 'if (beat.tags == null) { '
+      + '  beat.tags = []; '
+      + '} '
+      + 'if (!beat.tags.contains(params.tag)) { '
+      + '  beat.tags.add(params.tag); '
+      + '}';
+
+    return [
+      { update: { _id: `beat:${beatId}` } },
+      { script: { source: script, params: { tag } } }
+    ];
+  }));
+
+  const params = {
+    index: INDEX_NAMES.BEATS,
+    type: '_doc',
+    body,
+    refresh: 'wait_for'
+  };
+
+  const response = await callWithRequest('bulk', params);
+  return get(response, 'items', [])
+    .map((item, resultIdx) => ({
+      status: item.update.status,
+      result: item.update.result,
+      idxInRequest: assignments[resultIdx].idxInRequest
+    }));
+}
+
+function addNonExistentItemAssignmentsToResponse(response, assignments, nonExistentBeatIds, nonExistentTags) {
+  assignments.forEach((assignment, idx) => {
```

Comment:
We can make this `forEach` body cleaner if we destructure `tag` and `beat_id` out of `assignment` in the parameter declaration, since I think they're the only props needed.

```
assignments.forEach(({ beat_id, tag }, idx) => { ... })
```